server/container/guice/cassandra-guice/src/test/java/org/apache/james/FixingGhostMailboxTest.java

```diff
@@ -0,0 +1,318 @@
+        // State before ghost mailbox bug
+        // Alice INBOX is delegated to Bob and contains one message
+        aliceInboxPath = MailboxPath.forUser(alice, MailboxConstants.INBOX);
+        aliceInboxId = mailboxProbe.createMailbox(MailboxConstants.USER_NAMESPACE, alice, MailboxConstants.INBOX);
+        aclProbe.addRights(aliceInboxPath, bob, MailboxACL.FULL_RIGHTS);
+        message1 = mailboxProbe.appendMessage(alice, aliceInboxPath,
+            generateMessageContent(), new Date(), !RECENT, new Flags());
+        rule.await();
+
+        // Simulate ghost mailbox bug
+        session.execute(delete().from(CassandraMailboxPathTable.TABLE_NAME)
+            .where(eq(CassandraMailboxPathTable.NAMESPACE_AND_USER, mailboxBaseTupleUtil.createMailboxBaseUDT(MailboxConstants.USER_NAMESPACE, alice)))
+            .and(eq(CassandraMailboxPathTable.MAILBOX_NAME, MailboxConstants.INBOX)));
+
+        // trigger provisioning
+        given()
+            .header("Authorization", accessToken.serialize())
+            .body("[[\"getMailboxes\", {}, \"#0\"]]")
+        .when()
+            .post("/jmap")
+        .then()
+            .statusCode(200);
+
+        // Received a new message
+        message2 = mailboxProbe.appendMessage(alice, aliceInboxPath,
+            generateMessageContent(), new Date(), !RECENT, new Flags());
+        rule.await();
+    }
+
+    private ByteArrayInputStream generateMessageContent() {
+        return new ByteArrayInputStream("Subject: toto\r\n\r\ncontent".getBytes(StandardCharsets.UTF_8));
+    }
+
+    private URIBuilder baseUri() {
+        return new URIBuilder()
+            .setScheme("http")
+            .setHost("localhost")
+            .setPort(jmapServer.getProbe(JmapGuiceProbe.class)
+                .getJmapPort())
+            .setCharset(Charsets.UTF_8);
+    }
+
+    @After
+    public void teardown() {
+        jmapServer.stop();
+    }
+
+    @Test
+    public void ghostMailboxBugShouldChangeMailboxId() throws Exception {
+        Mailbox newAliceInbox = mailboxProbe.getMailbox(MailboxConstants.USER_NAMESPACE, alice, MailboxConstants.INBOX);
+
+        assertThat(aliceInboxId).isNotEqualTo(newAliceInbox.getMailboxId());
+    }
+
+    @Test
+    public void ghostMailboxBugShouldDiscardOldContent() throws Exception {
+        Mailbox newAliceInbox = mailboxProbe.getMailbox(MailboxConstants.USER_NAMESPACE, alice, MailboxConstants.INBOX);
+
+        given()
+            .header("Authorization", accessToken.serialize())
+            .body("[[\"getMessageList\", {\"filter\":{\"inMailboxes\":[\"" + newAliceInbox.getMailboxId().serialize() + "\"]}}, \"#0\"]]")
+        .when()
+            .post("/jmap")
+        .then()
+            .statusCode(200)
+            .body(NAME, equalTo("messageList"))
+            .body(ARGUMENTS + ".messageIds", hasSize(1))
+            .body(ARGUMENTS + ".messageIds", contains(message2.getMessageId().serialize()));
```

Comment:
Add ``.body ...... notContains(message1)`` will make test intention easier to understand IMO